### PR TITLE
[Feature]  Add ChamferDistance op in gpu

### PR DESCRIPTION
**Pull request by zytx121**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add ChamferDistance op, which is required by the paper `Oriented RepPoints for Aerial Object Detection (CVPR 2022)` in [MMRotate](https://github.com/open-mmlab/mmrotate/issues/127). 

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmcv/ops/chamfer_distance.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        torch.cuda.set_device(device)
+
+        ext_module.chamfer_distance_forward(xyz1, xyz2, dist1, dist2, idx1,
+                                            idx2)
+        ctx.save_for_backward(xyz1, xyz2, idx1, idx2)
+        return dist1, dist2, idx1, idx2
+
+    @staticmethod
+    @once_differentiable
+    def backward(ctx, graddist1, graddist2, gradidx1, gradidx2):
```
Comment (zhouzaida) [inline mmcv/ops/chamfer_distance.py:62]:
```suggestion
    def backward(ctx, grad_dist1, grad_dist2, grad_idx1, grad_idx2):
```

**Comment by grimoire** [comment]
Unit test failed with:

```bash
E       AssertionError: assert False
E        +  where False = <built-in method allclose of type object at 0x7fe58c83ccc0>(tensor([[0.0900, 0.4900, 0.4900, 0.0900],\n        [0.0900, 0.4900, 0.4900, 0.0900],\n        [0.5200, 0.6500, 0.4900, 0.3600]], device='cuda:0',\n       grad_fn=<ChamferDistanceFunctionBackward>), tensor([[0.5800, 0.4900, 0.1800, 0.4900],\n        [0.5800, 0.5200, 1.1600, 0.3600],\n        [0.0000, 0.0000, 0.0000, 0.0000]], device='cuda:0'), 0.01)
E        +    where <built-in method allclose of type object at 0x7fe58c83ccc0> = torch.allclose


```

**Comment by zhouzaida** [comment]
Hi @zytx121 , please update your branch with mmcv master which added CircleCI to support running GPU unit tests.

**Comment by zytx121** [comment]
> Unit test failed with:
> 
> ```shell
> E       AssertionError: assert False
> E        +  where False = <built-in method allclose of type object at 0x7fe58c83ccc0>(tensor([[0.0900, 0.4900, 0.4900, 0.0900],\n        [0.0900, 0.4900, 0.4900, 0.0900],\n        [0.5200, 0.6500, 0.4900, 0.3600]], device='cuda:0',\n       grad_fn=<ChamferDistanceFunctionBackward>), tensor([[0.5800, 0.4900, 0.1800, 0.4900],\n        [0.5800, 0.5200, 1.1600, 0.3600],\n        [0.0000, 0.0000, 0.0000, 0.0000]], device='cuda:0'), 0.01)
> E        +    where <built-in method allclose of type object at 0x7fe58c83ccc0> = torch.allclose
> ```

fixed

**Comment by zytx121** [comment]
> Hi @zytx121 , please update your branch with mmcv master which added CircleCI to support running GPU unit tests.

I have used ``git pull upstream master``, but it still doesn't run CircleCI.

File: mmcv/ops/chamfer_distance.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        dist2 = torch.zeros(batchsize, m)
+
+        idx1 = torch.zeros(batchsize, n).type(torch.IntTensor)
+        idx2 = torch.zeros(batchsize, m).type(torch.IntTensor)
+
+        dist1 = dist1.to(device)
+        dist2 = dist2.to(device)
+        idx1 = idx1.to(device)
+        idx2 = idx2.to(device)
+        torch.cuda.set_device(device)
```
Comment (grimoire) [inline mmcv/ops/chamfer_distance.py:53]: Can we remove `set_device`? We already have device guard, which is an RAII `set_device`.

File: mmcv/ops/chamfer_distance.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        return grad_xyz1, grad_xyz2
+
+
+class ChamferDistance(nn.Module):
+    """This is an implementation of the 2D Chamfer Distance."""
+
+    def forward(self, input1, input2):
+        """
+        Args:
+            input1 (torch.Tensor): Point set with shape (B, N, 2).
```
Comment (grimoire) [inline mmcv/ops/chamfer_distance.py:105]: If size of inputs can be different, should we use `(B, M, 2)` here?

File: mmcv/ops/csrc/common/cuda/chamfer_distance_cuda_kernel.cuh [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+                                                     int* result_i) {
+  __shared__ scalar_t buf[THREADS_PER_BLOCK * 2];
+  for (int i = blockIdx.x; i < b; i += gridDim.x) {
+    for (int k2 = 0; k2 < m; k2 += THREADS_PER_BLOCK) {
+      int end_k = min(m, k2 + THREADS_PER_BLOCK) - k2;
+      for (int j = threadIdx.x; j < end_k * 2; j += blockDim.x) {
+        buf[j] = xyz2[(i * m + k2) * 2 + j];
+      }
+      __syncthreads();
+      for (int j = threadIdx.x + blockIdx.y * blockDim.x; j < n;
```
Comment (grimoire) [inline mmcv/ops/csrc/common/cuda/chamfer_distance_cuda_kernel.cuh:27]: `blockIdx.y` is always 0?

File: mmcv/ops/csrc/common/cuda/chamfer_distance_cuda_kernel.cuh [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+           j += blockDim.x * gridDim.y) {
+        scalar_t x1 = xyz[(i * n + j) * 2 + 0];
+        scalar_t y1 = xyz[(i * n + j) * 2 + 1];
+        int best_i = 0;
+        scalar_t best = 0;
+        int end_ka = end_k - (end_k & 2);
```
Comment (grimoire) [inline mmcv/ops/csrc/common/cuda/chamfer_distance_cuda_kernel.cuh:33]: Use `int end_ka = endk&(~2);`

File: mmcv/ops/csrc/common/cuda/chamfer_distance_cuda_kernel.cuh [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+        if (end_ka == THREADS_PER_BLOCK) {
+          for (int k = 0; k < THREADS_PER_BLOCK; k += 4) {
+            {
```
Comment (grimoire) [inline mmcv/ops/csrc/common/cuda/chamfer_distance_cuda_kernel.cuh:36]: best can be initialized as a large number so we can remove `k==0` check.
You can use a loop with `#progma unoll` on top of it. Which will make the code more reaedable without lossing performance.

File: mmcv/ops/csrc/common/cuda/chamfer_distance_cuda_kernel.cuh [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+#include "pytorch_cuda_helper.hpp"
+#endif
+
+template <typename scalar_t>
+__global__ void chamfer_distance_forward_cuda_kernel(int b, int n,
+                                                     const scalar_t* xyz, int m,
+                                                     const scalar_t* xyz2,
+                                                     scalar_t* result,
+                                                     int* result_i) {
+  __shared__ scalar_t buf[THREADS_PER_BLOCK * 2];
```
Comment (grimoire) [inline mmcv/ops/csrc/common/cuda/chamfer_distance_cuda_kernel.cuh:19]: Could use a larger shared memory size according to the input shape.
Larger shared memory means less syncthread and less dram transfer.

File: mmcv/ops/chamfer_distance.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+            input2 (torch.Tensor): Point set with shape (B, N, 2).
+        """
+        input1 = input1.contiguous()
+        input2 = input2.contiguous()
```
Comment (zhouzaida) [inline mmcv/ops/chamfer_distance.py:109]: Can we move these two lines to `ChamferDistanceFunction.forward` and we can make `chamfer_distance = ChamferDistanceFunction.apply`?

Comment (zhouzaida) [inline mmcv/ops/chamfer_distance.py:109]: Can we move these two lines to `ChamferDistanceFunction.forward` and we can make `chamfer_distance = ChamferDistanceFunction.apply`?

File: mmcv/ops/chamfer_distance.py (hunk 2 above)
Comment (zytx121) [reply mmcv/ops/chamfer_distance.py:53]: ok

File: mmcv/ops/chamfer_distance.py (hunk 8 above)
Comment (zytx121) [reply mmcv/ops/chamfer_distance.py:109]: Good  suggestion！

File: mmcv/ops/csrc/common/cuda/chamfer_distance_cuda_kernel.cuh (hunk 6 above)
Comment (zytx121) [reply mmcv/ops/csrc/common/cuda/chamfer_distance_cuda_kernel.cuh:36]: Get!

**Comment by zytx121** [comment]
Kindly ping @grimoire to have a look again.

File: mmcv/ops/csrc/common/cuda/chamfer_distance_cuda_kernel.cuh [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+      }
+      __syncthreads();
+      for (int j = threadIdx.x; j < n; j += blockDim.x * gridDim.y) {
+        scalar_t x1 = xyz[(i * n + j) * 2 + 0];
+        scalar_t y1 = xyz[(i * n + j) * 2 + 1];
+        int best_i = 0;
+        scalar_t best = 1e10;
+        int end_ka = end_k & (~2);
+#progma unroll
+        for (int k = 0; k < end_ka; k) {
```
Comment (grimoire) [inline mmcv/ops/csrc/common/cuda/chamfer_distance_cuda_kernel.cuh:34]: Err, end_ka is not a const value so the loop can not been unrolled.
You still need ` if (end_ka == THREADS_PER_BLOCK)` and the loop would look like :
```c++
for(int k=0;k< end_ka; k+=4){
    #pragma unroll
    for(int j=0;j<4;++j){
        ....
    }
}
```

**Review by grimoire — APPROVED** [review]
LGTM

File: mmcv/ops/chamfer_distance.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+                - dist1 (Tensor): Chamfer ditacne (xyz1 to xyz2) with
+                    shape (B, N).
+                - dist2 (Tensor): Chamfer ditacne (xyz2 to xyz1) with
+                    shape (B, N).
+                - idx1 (Tensor): Index of chamfer ditacne (xyz1 to xyz2)
+                    with shape (B, N), which be used in compute gradient.
+                - idx2 (Tensor): Index of chamfer ditacne (xyz2 to xyz2)
+                    with shape (B, N), which be used in compute gradient.
+        """
+        batchsize, n, _ = xyz1.size()
```
Comment (zhouzaida) [inline mmcv/ops/chamfer_distance.py:41]:
```suggestion
        batch_size, n, _ = xyz1.size()
```

File: mmcv/ops/chamfer_distance.py [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+
+        Args:
+            grad_dist1 (Tensor): Gradient of chamfer ditacne
+                (xyz1 to xyz2) with shape (B, N).
+            grad_dist2 (Tensor): Gradient of chamfer ditacne
+                (xyz2 to xyz1) with shape (B, N).
+            grad_idx1 (Tensor): Index of chamfer ditacne (xyz1 to xyz2)
+                    with shape (B, N), which be used in compute gradient.
+            grad_idx2 (Tensor): Index of chamfer ditacne (xyz2 to xyz2)
+                    with shape (B, N), which be used in compute gradient.
```
Comment (zhouzaida) [inline mmcv/ops/chamfer_distance.py:72]:
```suggestion
            grad_idx1 (Tensor): Index of chamfer ditacne (xyz1 to xyz2)
                with shape (B, N), which be used in compute gradient.
            grad_idx2 (Tensor): Index of chamfer ditacne (xyz2 to xyz2)
                with shape (B, N), which be used in compute gradient.
```

**Review by zhouzaida — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**